XMonad/Hooks/DynamicStatusBarConfigs.hs

```diff
@@ -0,0 +1,125 @@
+{-# LANGUAGE TupleSections #-}
+-----------------------------------------------------------------------------
+-- |
+-- Module      :  XMonad.Hooks.DynamicStatusBarConfigs
+-- Copyright   :  (c) Yecine Megdiche <yecine.megdiche@gmail.com>
+-- License     :  BSD3-style (see LICENSE)
+--
+-- Maintainer  :  Yecine Megdiche <yecine.megdiche@gmail.com>
+-- Stability   :  unstable
+-- Portability :  unportable
+--
+-- This module provides dynamic status bars, that react to screen changes.
+-- Heavily inspired by "XMonad.Hooks.DynamicBars".
+----------------------------------------------------------------------------
+module XMonad.Hooks.DynamicStatusBarConfigs
+  (
+    -- * Usage
+    -- $usage
+    dynamicSBs
+  )
+where
+
+
+import           Data.List                      ( partition
+                                                , (\\)
+                                                )
+import           Data.Foldable                  ( traverse_ )
+import           Data.Monoid                    ( All(..) )
+import           Graphics.X11.Xrandr            ( xrrSelectInput )
+
+import           XMonad
+import           XMonad.Hooks.DynamicLog
+import qualified XMonad.StackSet               as W
+import qualified XMonad.Util.ExtensibleState   as XS
+
+-- $usage
+-- Using multiple status bars just with "XMonad.Hooks.DynamicLog' works well
+-- as long as the screen configuration does not change often. Using
+-- 'dynamicSBs' fixes that problem: by providing a function that creates
+-- status bars, it takes care of setting up the event hook, the log hook
+-- and the startup hook necessary to make the status bars, well, dynamic.
+--
+-- > mainTopBar, mainBottomBar, secondaryBar1 :: IO StatusBarConfig
+-- > mainTopBar = statusBarPropConfig "xmobar -x 0 $HOME/.config/xmobar/xmobarrc_top" ppTop
+-- > mainBottomBar = statusBarHandleConfig "xmobar -x 0 $HOME/.config/xmobar/xmobarrc_bottom" ppBottom
+-- > secondaryBar1 = statusBarPropConfig "xmobar -x 1 $HOME/.config/xmobar/xmobarrc1" pp1
+-- >
+-- > barSpawner :: ScreenId -> IO StatusBarConfig
+-- > barSpawner 0 = mainTopBar <> mainBottomBar -- two bars on the main screen
+-- > barSpawner 1 = secondaryBar1
+-- > barSpawner _ = memtpy -- nothing on the rest of the screens
+-- >
+-- > main = xmonad $ dynamicSBs barSpawner (def { ... })
+
+newtype ActiveSBs = ASB {getASBs :: [(ScreenId,  StatusBarConfig)]}
+
+instance ExtensionClass ActiveSBs where
+  initialValue = ASB []
+
+-- | Given a function to create status bars, 'dynamicSBs'
+-- adds the dynamic status bar capabilities to the config.
+dynamicSBs :: (ScreenId -> IO StatusBarConfig) -> XConfig l -> XConfig l
+dynamicSBs f conf = conf
+  { startupHook     = startupHook conf
+                      *> setupEventHandler
+                      *> cleanAll
+                      *> updateSBs f
+  , logHook         = logHook conf *> logSBs
+  , handleEventHook = eventHookSBs f <> handleEventHook conf
```

Comment:
I could maybe synthesize the RRScreenChangeNotify event that was discarded by Rescreen and explicitly invoke handleEventHook on it. It's a horrible hack, though.

Reply:
I only took that event because it was the one used in `DynamicBars`, but I don't really know if it's the best choice...

What's the problem here?

Reply:
The problem is that my `X.H.Rescreen` module completely replaces xmonad's handling of Configure and RRScreenChangeNotify events in a way that doesn't compose well with other hooks (prevents other hooks from seeing all the events, prevent xmonad core from handling any of the events). It needs to do this because the X server sends several such events on every xrandr change, so

1. if one wants to attach a hook that restarts something (bars that don't reposition themselves, for example), the events need to be debounced for the hook to be invoked only once
2. if one wants to attach two different hooks, one for actual screen layout reconfiguration, another for just connect/disconnect of outputs, we need to look at all the events to detect which of the two situations it was

If this went into the core with two additional hooks in XConfig, everything would then compose nicely, but the xmonad philosophy (as I see it) has always been to keep the core minimal and try to do as much as possible in optional contrib modules. It usually works well, it just sometimes gets awkward if contirb modules interact/overlap as we never know if the user will use one, the other, or both.

Reply:
I'm not sure I follow, don't all event handlers that are composed with `<>` see the same events? I think the `All False` will only prevent the xmonad default handler from being dispatched, and it doesn't affect the other events right?

Reply:
Correct, you would need to add your own logic to suppress your own event handlers in that case. Note that xmonad has no idea what hooks return `All False` until the hook has executed, since they can contain arbitrary logic.

Reply:
All events handlers do see the same events, but `X.H.Rescreen` looks into the X event queue directly and discard duplicate events, so other hooks only see the first Configure or RRScreenChangeNotify event, whichever comes first, and all the other events are discarded. I should probably explain this more clearly in the documentation there.

Reply:
Alright that makes sense now

Well, I guess we should export the event handler (and the other hooks) from this module and explain how to make it work with `X.H.Rescreen` (Adding it to the `RescreenConfig`, and manually add the startup hook). I'm not sure that calling the `handleEvent` on it manually is a good idea

Reply:
@liskin  so what should we do here?

Reply:
Nothing for now. :-)

I got an idea that might improve composability of stuff like this (something like extensible state, just for XConfig, an "extensible config"). But I'd like to finish other stuff first, so let's not worry about the interaction here and deal with it later. I don't really need X.H.Rescreen in my config, auto-xrandr is more trouble than it's worth, so I'll just try using your code and see how it behaves. (But I won't get to it sooner than end of this week.)

Reply:
This is not at all urgent, so take your time!